File: server/encoder.py
from server.config import (
    ACTIVE_CHANNELS,
    CHANNEL_TO_POS,
    GRID_SIZE,
    STIM_CURRENT_UA,
    STIM_DURATION_US,
)


def compute_stim_channels(ball_x: float, ball_y: float) -> list[int]:
    """Return which MEA channels fall within stimulation radius of ball position."""
    target_row = int(ball_y * (GRID_SIZE - 1))
    target_col = int(ball_x * (GRID_SIZE - 1))

    channels = []
    for ch in ACTIVE_CHANNELS:
        row, col = CHANNEL_TO_POS[ch]
        dist = abs(row - target_row) + abs(col - target_col)
        if dist <= 2:
            channels.append(ch)
    return channels


def encode_and_stim(neurons, ball_x: float, ball_y: float) -> list[int]:
    """Compute stim channels and deliver stimulation via CL SDK."""
    channels = compute_stim_channels(ball_x, ball_y)
    if not channels:
        return []

    import cl

    channel_set = cl.ChannelSet(*channels)
    design = cl.StimDesign(
        STIM_DURATION_US, -STIM_CURRENT_UA, STIM_DURATION_US, STIM_CURRENT_UA
    )
    neurons.stim(channel_set, design)
    return channels

File: web/app/engine/encoder.ts
import { ACTIVE_CHANNELS, CHANNEL_TO_POS, GRID_SIZE } from "./config";

/**
 * Compute organized stim channels — ball position maps to nearby electrodes.
 * This is the "reward" signal: predictable, structured input.
 */
export function computeStimChannels(ballX: number, ballY: number): number[] {
  const targetRow = Math.floor(ballY * (GRID_SIZE - 1));
  const targetCol = Math.floor(ballX * (GRID_SIZE - 1));

  const channels: number[] = [];
  for (const ch of ACTIVE_CHANNELS) {
    const pos = CHANNEL_TO_POS.get(ch)!;
    const dist = Math.abs(pos[0] - targetRow) + Math.abs(pos[1] - targetCol);
    if (dist <= 2) channels.push(ch);
  }
  return channels;
}

/**
 * Compute chaotic stim channels — random subset of electrodes.
 * This is the "punishment" signal: unpredictable noise,
 * inspired by DishBrain's Free Energy Principle approach.
 */
export function computeChaoticStimChannels(): number[] {
  const count = 5 + Math.floor(Math.random() * 10);
  const shuffled = [...ACTIVE_CHANNELS].sort(() => Math.random() - 0.5);
  return shuffled.slice(0, count);
}
